server.js

```diff
@@ -2,29 +2,80 @@ import express from 'express'
 import bodyParser from 'body-parser'
 import cors from 'cors'
 
-// If you're using one of our datasets, uncomment the appropriate import below
-// to get started!
-// 
-// import goldenGlobesData from './data/golden-globes.json'
-// import avocadoSalesData from './data/avocado-sales.json'
-// import booksData from './data/books.json'
-// import netflixData from './data/netflix-titles.json'
-// import topMusicData from './data/top-music.json'
-
-// Defines the port the app will run on. Defaults to 8080, but can be 
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
+import harryData from './data/harry-potter-characters.json'
+
+const ERROR_CHARACTERS_NOT_FOUND = {error : 'No character results were found, please try again.'}
+
+// Defines the port the app will run on
 const port = process.env.PORT || 8080
 const app = express()
 
-// Add middlewares to enable cors and json body parsing
+// Middlewares to enable cors and json body parsing
 app.use(cors())
 app.use(bodyParser.json())
 
-// Start defining your routes here
-app.get('/', (req, res) => {
-  res.send('Hello world')
+// Starting route
+app.get('/', (request, response) => {
+  response.send('Hello, this is an API with Harry Potter characters. Go to https://lindas-project-express-api.herokuapp.com/characters to see all the characters. For more documentation, visit: https://github.com/lindahz/project-express-api')
+})
+
+// This endpoint shows all characters + will let you filter on name with query parameter
+app.get('/characters', (request, response) => {
+  const { name } = request.query
+  const filteredName = harryData.filter(character => 
+    character.name.toLocaleLowerCase() && 
+    character.name.toLocaleLowerCase().includes(name))
+
+  if (filteredName.length === 0) {
+    response.json(harryData)
+  } else {
+    response.json(filteredName)
+  }
+})
+
+// This endpoint shows one character from a unique id
+app.get('/characters/id/:id', (request, response) => {
+  const { id } = request.params
+  const showId = harryData.find((item) => item.id === +id)
+  // Include error response here
+  if (!showId) {
+    response.status(404).json(ERROR_CHARACTERS_NOT_FOUND)
+  } else {
+    response.json(showId)
+  }
+})
+
+// This enpoint shows all names of the characters, sorted from A-Z
+app.get('/characters/names', (request, response) => {
+  const names = harryData.map(name => name.name).sort()
+  response.json(names)
+})
+
+// This endpoint shows all characters from a specific house
+app.get('/characters/:house', (request, response) => {
+  const { house } = request.params
+  const houses = harryData.filter((item) => item.house.toLocaleLowerCase() === house)
+
+  if (houses.length === 0) {
+    response.status(404).json(ERROR_CHARACTERS_NOT_FOUND)
```

Comment:
Once again, we should return empty array rather than error. Think about it like this - when we query for multiple elements, we should ALWAYS return array. It's either going to be array of elements (if we fulfilled queries) or just empty array.
We should return 404 error when we query for single element. Only when we queried for single element and that query was not fulfilled (for example looking for house "Technigo"), then we should return 404 error

Reply:
*** Needs to be fixed ***